src/ReinforcementLearningCore/src/policies/agent/agent_base.jl

```diff
@@ -51,39 +49,37 @@ function RLBase.optimise!(policy::AbstractPolicy, stage::AbstractStage, trajecto
 
 @functor Agent (policy,)
 
-function Base.push!(agent::Agent, ::PreActStage, env::AbstractEnv)
-    push!(agent, state(env))
+function Base.push!(agent::Agent, ::PreEpisodeStage, env::AbstractEnv)
+    push!(agent.trajectory, (state = state(env),))
 end
 
 # !!! TODO: In async scenarios, parameters of the policy may still be updating
 # (partially), which will result to incorrect action. This should be addressed
 # in Oolong.jl with a wrapper
-function RLBase.plan!(agent::Agent{P,T,C}, env::AbstractEnv) where {P,T,C}
-    action = RLBase.plan!(agent.policy, env)
-    push!(agent.trajectory, agent.cache, action)
-    action
+function RLBase.plan!(agent::Agent, env::AbstractEnv)
+    RLBase.plan!(agent.policy, env)
 end
 
 # Multiagent Version
-function RLBase.plan!(agent::Agent{P,T,C}, env::E, p::Symbol) where {P,T,C,E<:AbstractEnv}
-    action = RLBase.plan!(agent.policy, env, p)
-    push!(agent.trajectory, agent.cache, action)
-    action
+function RLBase.plan!(agent::Agent, env::AbstractEnv, p::Symbol)
+    RLBase.plan!(agent.policy, env, p)
 end
 
-function Base.push!(agent::Agent{P,T,C}, ::PostActStage, env::E) where {P,T,C,E<:AbstractEnv}
-    push!(agent.cache, reward(env), is_terminated(env))
+function Base.push!(agent::Agent, ::PostActStage, env::AbstractEnv, action)
+    next_state = state(env)
+    push!(agent.trajectory, (state = next_state, action = action, reward = reward(env), terminal = is_terminated(env)))
 end
 
-function Base.push!(agent::Agent, ::PostExperimentStage, env::E) where {E<:AbstractEnv}
-    RLBase.reset!(agent.cache)
-end
-
-function Base.push!(agent::Agent, ::PostExperimentStage, env::E, player::Symbol) where {E<:AbstractEnv}
-    RLBase.reset!(agent.cache)
+function Base.push!(agent::Agent, ::PostEpisodeStage, env::E)
+    if haskey(agent.trajectory, :next_action) 
```

Comment:
This is if the episode finished (whether truncated or terminated) we query the policy to plan another step. We should also check if the environment is not terminated? If it is, it just makes no sense to plan an action.

Reply:
It wouldn't make sense indeed, but if your environment has terminal states at all, then you should not use a trajectory that has a next_action key. That's the how I thought about it. If we add that check, then it allows the user to have an incorrect trajectory without an error being thrown and the buffer will accumulate mistakes.